File: src/data/sensors/TelemetrySimulator.ts
import type { DrivingState, TelemetryPoint } from '../../domain/models/telemetry';

export interface SimulationPreset {
  id: string;
  name: string;
  description: string;
  baseLat: number;
  baseLng: number;
  targetSpeedKmH: number;
  trafficProbability: number;
}

export const SIMULATION_PRESETS: SimulationPreset[] = [
  {
    id: 'preset-urban',
    name: 'Sri Lanka Coastal Drive (Galle Road A2)',
    description: 'Real-time Sri Lankan road telemetry with signal holds and speed variance.',
    baseLat: 6.7130,
    baseLng: 79.9026,
    targetSpeedKmH: 45,
    trafficProbability: 0.20,
  },
];

function haversineKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

export class TelemetrySimulator {
  private timerId: any = null;
  private isRunning = false;
  private currentPreset: SimulationPreset = SIMULATION_PRESETS[0];
  private callback: ((point: TelemetryPoint) => void) | null = null;

  private stepIndex = 0;
  private lat: number = SIMULATION_PRESETS[0].baseLat;
  private lng: number = SIMULATION_PRESETS[0].baseLng;
  private customStartLat: number | null = null;
  private customStartLng: number | null = null;

  private destLat: number | null = null;
  private destLng: number | null = null;
  private routePath: [number, number][] = [];
  private pathIdx = 0;

  private currentSpeed = 0;
  private tripId = `trip-${Date.now()}`;

  // Manual event overrides
  private pendingEvent: 'HARD_ACCEL' | 'HARD_BRAKE' | 'SWERVE' | 'TRAFFIC_JAM' | 'PIT_STOP' | null = null;
  private trafficJamTimer = 0;
  private pitStopTimer = 0;

  public setPreset(presetId: string): void {
    const found = SIMULATION_PRESETS.find((p) => p.id === presetId);
    if (found) {
      this.currentPreset = found;
    }
  }

  public setStartPosition(lat: number, lng: number): void {
    this.customStartLat = lat;
    this.customStartLng = lng;
    this.lat = lat;
    this.lng = lng;
  }

  public setDestination(destLat: number | null, destLng: number | null, path: [number, number][] = []): void {
    this.destLat = destLat;
    this.destLng = destLng;
    this.routePath = path;
    this.pathIdx = 0;
  }

  public triggerManualEvent(event: 'HARD_ACCEL' | 'HARD_BRAKE' | 'SWERVE' | 'TRAFFIC_JAM' | 'PIT_STOP'): void {
    this.pendingEvent = event;
    if (event === 'TRAFFIC_JAM') this.trafficJamTimer = 18;
    if (event === 'PIT_STOP') this.pitStopTimer = 125;
  }

  public start(onPoint: (point: TelemetryPoint) => void): void {
    if (this.isRunning) return;
    this.isRunning = true;
    this.callback = onPoint;
    this.stepIndex = 0;
    this.pathIdx = 0;
    this.tripId = `trip-${Date.now()}`;

    if (this.customStartLat !== null && this.customStartLng !== null) {
      this.lat = this.customStartLat;
      this.lng = this.customStartLng;
    } else {
      this.lat = this.currentPreset.baseLat;
      this.lng = this.currentPreset.baseLng;
    }

    this.currentSpeed = 10;
    this.timerId = setInterval(() => this.tick(), 1000);
  }

  public stop(): void {
    if (this.timerId) {
      clearInterval(this.timerId);
      this.timerId = null;
    }
    this.isRunning = false;
  }

  public isActive(): boolean {
    return this.isRunning;
  }

  public getTripId(): string {
    return this.tripId;
  }

  private tick(): void {
    if (!this.callback) return;

    this.stepIndex++;
    const now = Date.now();

    // 1. If following a road polyline path, step strictly along the road path!
    if (this.routePath.length > 0) {
      if (this.pathIdx < this.routePath.length) {
        const pt = this.routePath[this.pathIdx];
        this.lat = pt[0];
        this.lng = pt[1];
        this.pathIdx += Math.max(1, Math.floor(this.currentSpeed / 25));

        if (this.pathIdx >= this.routePath.length) {
          // Reached Destination! Stop cleanly.
          const finalPt = this.routePath[this.routePath.length - 1];
          this.lat = finalPt[0];
          this.lng = finalPt[1];
          this.currentSpeed = 0;
          this.emitPoint(now, 0, 9.81, 0, 0, 'OFFLINE');
          this.stop();
          return;
        }
      }
    } else if (this.destLat !== null && this.destLng !== null) {
      // Check distance to destination
      const remainingKm = haversineKm(this.lat, this.lng, this.destLat, this.destLng);
      if (remainingKm <= 0.08) {
        this.lat = this.destLat;
        this.lng = this.destLng;
        this.currentSpeed = 0;
        this.emitPoint(now, 0, 9.81, 0, 0, 'OFFLINE');
        this.stop();
        return;
      }
    }

    // 2. Handle Active Idling / Pit-Stop Timers
    if (this.pitStopTimer > 0) {
      this.pitStopTimer--;
      this.currentSpeed = 0;
      this.emitPoint(now, 0, 9.81, 0, 0, 'PIT_STOP');
      return;
    }

    if (this.trafficJamTimer > 0) {
      this.trafficJamTimer--;
      this.currentSpeed = 0.5;
      this.emitPoint(now, 0.1, 9.81, -0.2, 0, 'IDLING');
      return;
    }

    // 3. Handle One-Shot Manual Events
    if (this.pendingEvent === 'HARD_ACCEL') {
      this.pendingEvent = null;
      this.currentSpeed += 25;
      this.emitPoint(now, 0.5, 9.81, 4.2, 5, 'DRIVING', 'HARD_ACCEL');
      return;
    }

    if (this.pendingEvent === 'HARD_BRAKE') {
      this.pendingEvent = null;
      this.currentSpeed = Math.max(5, this.currentSpeed - 30);
      this.emitPoint(now, -0.8, 9.81, -4.8, -8, 'DRIVING', 'HARD_BRAKE');
      return;
    }

    if (this.pendingEvent === 'SWERVE') {
      this.pendingEvent = null;
      this.emitPoint(now, 2.8, 9.81, 0.4, 54.0, 'DRIVING', 'SWERVE');
      return;
    }

    // 4. Dynamic Speed Control
    const targetSpeed = this.currentPreset.targetSpeedKmH;
    const speedNoise = (Math.random() - 0.5) * 6;
    this.currentSpeed += (targetSpeed - this.currentSpeed) * 0.1 + speedNoise;
    this.currentSpeed = Math.max(0, Math.min(120, this.currentSpeed));

    if (this.routePath.length === 0) {
      let dirLat = -0.0003;
      let dirLng = 0.0001;

      if (this.destLat !== null && this.destLng !== null) {
        const dLat = this.destLat - this.lat;
        const dLng = this.destLng - this.lng;
        const dist = Math.sqrt(dLat * dLat + dLng * dLng);

        if (dist > 0.0001) {
          const stepSize = (this.currentSpeed / 360000);
          dirLat = (dLat / dist) * stepSize;
          dirLng = (dLng / dist) * stepSize;
        }
      }

      this.lat += dirLat;
      this.lng += dirLng;
    }

    const accelZ = parseFloat(((this.currentSpeed - targetSpeed) * 0.05 + (Math.random() - 0.5) * 0.8).toFixed(2));
    const gyroZ = parseFloat(((Math.random() - 0.5) * 12).toFixed(1));
    const state: DrivingState = this.currentSpeed < 3 ? 'IDLING' : 'DRIVING';

    this.emitPoint(now, 0, 9.81, accelZ, gyroZ, state);
  }

  private emitPoint(
    timestamp: number,
    accelX: number,
    accelY: number,
    accelZ: number,
    gyroZ: number,
    state: DrivingState,
    penalty?: TelemetryPoint['safetyPenalty']
  ): void {
    const point: TelemetryPoint = {
      id: `pt-${this.stepIndex}`,
      tripId: this.tripId,
      timestamp,
      latitude: parseFloat(this.lat.toFixed(6)),
      longitude: parseFloat(this.lng.toFixed(6)),
      speedKmH: parseFloat(this.currentSpeed.toFixed(1)),
      heading: 45,
      accelX,
      accelY,
      accelZ,
      gyroX: 0,
      gyroY: 0,
      gyroZ,
      state,
      gForceCombined: parseFloat((1.0 + Math.abs(accelZ) / 9.81).toFixed(2)),
      safetyPenalty: penalty,
    };

    if (this.callback) {
      this.callback(point);
    }
  }
}

export const telemetrySimulator = new TelemetrySimulator();

File: smart_trip_ai_mobile/lib/data/sensors/telemetry_simulator.dart
import 'dart:async';
import 'dart:math';
import '../../domain/models/telemetry.dart';

class SimulationPreset {
  final String id;
  final String name;
  final String description;
  final double baseLat;
  final double baseLng;
  final double targetSpeedKmH;
  final double trafficProbability;

  const SimulationPreset({
    required this.id,
    required this.name,
    required this.description,
    required this.baseLat,
    required this.baseLng,
    required this.targetSpeedKmH,
    required this.trafficProbability,
  });
}

const List<SimulationPreset> simulationPresets = [
  SimulationPreset(
    id: 'preset-urban',
    name: 'Sri Lanka Coastal Drive (Galle Road A2)',
    description: 'Real-time Sri Lankan road telemetry with signal holds and speed variance.',
    baseLat: 6.7130,
    baseLng: 79.9026,
    targetSpeedKmH: 45,
    trafficProbability: 0.20,
  ),
];

double _haversineKm(double lat1, double lon1, double lat2, double lon2) {
  const r = 6371.0;
  final dLat = ((lat2 - lat1) * pi) / 180;
  final dLon = ((lon2 - lon1) * pi) / 180;
  final a = sin(dLat / 2) * sin(dLat / 2) +
      cos((lat1 * pi) / 180) *
          cos((lat2 * pi) / 180) *
          sin(dLon / 2) *
          sin(dLon / 2);
  final c = 2 * atan2(sqrt(a), sqrt(1 - a));
  return r * c;
}

class TelemetrySimulator {
  Timer? _timer;
  bool isRunning = false;
  SimulationPreset currentPreset = simulationPresets[0];
  Function(TelemetryPoint)? callback;

  int stepIndex = 0;
  double lat = simulationPresets[0].baseLat;
  double lng = simulationPresets[0].baseLng;
  double? customStartLat;
  double? customStartLng;

  double? destLat;
  double? destLng;
  List<List<double>> routePath = [];
  int pathIdx = 0;

  double currentSpeed = 0;
  String tripId = 'trip-${DateTime.now().millisecondsSinceEpoch}';

  String? pendingEvent;
  int trafficJamTimer = 0;
  int pitStopTimer = 0;
  final _random = Random();

  void setPreset(String presetId) {
    try {
      currentPreset = simulationPresets.firstWhere((p) => p.id == presetId);
    } catch (_) {}
  }

  void setStartPosition(double newLat, double newLng) {
    customStartLat = newLat;
    customStartLng = newLng;
    lat = newLat;
    lng = newLng;
  }

  void setDestination(double? dLat, double? dLng, [List<List<double>> path = const []]) {
    destLat = dLat;
    destLng = dLng;
    routePath = path;
    pathIdx = 0;
  }

  void triggerManualEvent(String event) {
    pendingEvent = event;
    if (event == 'TRAFFIC_JAM') trafficJamTimer = 18;
    if (event == 'PIT_STOP') pitStopTimer = 125;
  }

  void start(Function(TelemetryPoint) onPoint) {
    if (isRunning) return;
    isRunning = true;
    callback = onPoint;
    stepIndex = 0;
    pathIdx = 0;
    tripId = 'trip-${DateTime.now().millisecondsSinceEpoch}';

    if (customStartLat != null && customStartLng != null) {
      lat = customStartLat!;
      lng = customStartLng!;
    } else {
      lat = currentPreset.baseLat;
      lng = currentPreset.baseLng;
    }

    currentSpeed = 10;
    _timer = Timer.periodic(const Duration(seconds: 1), (_) => _tick());
  }

  void stop() {
    _timer?.cancel();
    _timer = null;
    isRunning = false;
  }

  bool isActive() => isRunning;
  String getTripId() => tripId;

  void _tick() {
    if (callback == null) return;

    stepIndex++;
    final now = DateTime.now().millisecondsSinceEpoch;

    // 1. If following a road polyline path, step strictly along the road path!
    if (routePath.isNotEmpty) {
      if (pathIdx < routePath.length) {
        final pt = routePath[pathIdx];
        lat = pt[0];
        lng = pt[1];
        pathIdx += max(1, (currentSpeed / 25).floor());

        if (pathIdx >= routePath.length) {
          // Reached Destination! Stop cleanly.
          final finalPt = routePath.last;
          lat = finalPt[0];
          lng = finalPt[1];
          currentSpeed = 0;
          _emitPoint(now, 0, 9.81, 0, 0, DrivingState.offline);
          stop();
          return;
        }
      }
    } else if (destLat != null && destLng != null) {
      // Check distance to destination
      final remainingKm = _haversineKm(lat, lng, destLat!, destLng!);
      if (remainingKm <= 0.08) {
        lat = destLat!;
        lng = destLng!;
        currentSpeed = 0;
        _emitPoint(now, 0, 9.81, 0, 0, DrivingState.offline);
        stop();
        return;
      }
    }

    // 2. Handle Active Idling / Pit-Stop Timers
    if (pitStopTimer > 0) {
      pitStopTimer--;
      currentSpeed = 0;
      _emitPoint(now, 0, 9.81, 0, 0, DrivingState.pitStop);
      return;
    }

    if (trafficJamTimer > 0) {
      trafficJamTimer--;
      currentSpeed = 0.5;
      _emitPoint(now, 0.1, 9.81, -0.2, 0, DrivingState.idling);
      return;
    }

    // 3. Handle One-Shot Manual Events
    if (pendingEvent == 'HARD_ACCEL') {
      pendingEvent = null;
      currentSpeed += 25;
      _emitPoint(now, 0.5, 9.81, 4.2, 5, DrivingState.driving, 'HARD_ACCEL');
      return;
    }

    if (pendingEvent == 'HARD_BRAKE') {
      pendingEvent = null;
      currentSpeed = max(5, currentSpeed - 30);
      _emitPoint(now, -0.8, 9.81, -4.8, -8, DrivingState.driving, 'HARD_BRAKE');
      return;
    }

    if (pendingEvent == 'SWERVE') {
      pendingEvent = null;
      _emitPoint(now, 2.8, 9.81, 0.4, 54.0, DrivingState.driving, 'SWERVE');
      return;
    }

    // 4. Dynamic Speed Control
    final targetSpeed = currentPreset.targetSpeedKmH;
    final speedNoise = (_random.nextDouble() - 0.5) * 6;
    currentSpeed += (targetSpeed - currentSpeed) * 0.1 + speedNoise;
    currentSpeed = max(0, min(120, currentSpeed));

    if (routePath.isEmpty) {
      double dirLat = -0.0003;
      double dirLng = 0.0001;

      if (destLat != null && destLng != null) {
        final dLat = destLat! - lat;
        final dLng = destLng! - lng;
        final dist = sqrt(dLat * dLat + dLng * dLng);

        if (dist > 0.0001) {
          final stepSize = (currentSpeed / 360000);
          dirLat = (dLat / dist) * stepSize;
          dirLng = (dLng / dist) * stepSize;
        }
      }

      lat += dirLat;
      lng += dirLng;
    }

    final accelZValue = (currentSpeed - targetSpeed) * 0.05 + (_random.nextDouble() - 0.5) * 0.8;
    final accelZ = double.parse(accelZValue.toStringAsFixed(2));
    final gyroZValue = (_random.nextDouble() - 0.5) * 12;
    final gyroZ = double.parse(gyroZValue.toStringAsFixed(1));
    final state = currentSpeed < 3 ? DrivingState.idling : DrivingState.driving;

    _emitPoint(now, 0, 9.81, accelZ, gyroZ, state);
  }

  void _emitPoint(
    int timestamp,
    double accelX,
    double accelY,
    double accelZ,
    double gyroZ,
    DrivingState state,
    [String? penalty]
  ) {
    final point = TelemetryPoint(
      id: 'pt-$stepIndex',
      tripId: tripId,
      timestamp: timestamp,
      latitude: double.parse(lat.toStringAsFixed(6)),
      longitude: double.parse(lng.toStringAsFixed(6)),
      speedKmH: double.parse(currentSpeed.toStringAsFixed(1)),
      heading: 45,
      accelX: accelX,
      accelY: accelY,
      accelZ: accelZ,
      gyroX: 0,
      gyroY: 0,
      gyroZ: gyroZ,
      state: state,
      gForceCombined: double.parse((1.0 + accelZ.abs() / 9.81).toStringAsFixed(2)),
      safetyPenalty: penalty,
    );

    if (callback != null) {
      callback!(point);
    }
  }
}

final telemetrySimulator = TelemetrySimulator();
